File: 303.区域和检索-数组不可变.py
#
# @lc app=leetcode.cn id=303 lang=python3
#
# [303] 区域和检索 - 数组不可变
#

# @lc code=start
class NumArray:

    def __init__(self, nums: List[int]):
        self.sums=[0]
        # _sums=self.sums # _代表内部变量 无法修改，这里不一定要内部变量，普通的即可
        

        for num in nums:
            self.sums.append(self.sums[-1]+num) # 求和的每个元素
            
    def sumRange(self, left: int, right: int) -> int:
        return self.sums[right+1]-self.sums[left]



# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)
# @lc code=end

File: 303.区域和检索-数组不可变.cpp
/*
 * @lc app=leetcode.cn id=303 lang=cpp
 *
 * [303] 区域和检索 - 数组不可变
 */

// @lc code=start
class NumArray {
public:
    // 定义前缀和数组
    vector<int> sums;
    
    NumArray(vector<int>& nums) {
        int n=nums.size();
        sums.resize(n+1);
       
        for(int i=0;i<n;++i){
            // sums[ i ] 表示数组 nums 从下标 0 到下标 i - 1 的所有元素之和
            // 初始化 sums 数组
            sums[i+1]=sums[i]+nums[i];
        }
    }
    
    int sumRange(int left, int right) {
        return sums[right+1]-sums[left];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */
// @lc code=end

